### PR TITLE
[5.6][stdlib] Switch to new single-header modulemap for Android and OpenBSD too

### PR DESCRIPTION
**Pull request by finagolfin**
Cherrypick of #35707 and #40557
 
__Explanation:__ #32404 switched linux glibc over to a single-header modulemap last year, so Android and OpenBSD are following suit. Just like on linux, this fixes issues with C++ interoperability.

__Scope:__ Only affects these external platforms, with the exception of the three header changes in `SwiftGlibc.h.gyb`.

__SR Issue:__ None

__Risk:__ Very low

__Testing:__ I've been applying [the Android pull on my daily Android CI when cross-compiling the trunk and 5.6 stdlib](https://github.com/buttaface/swift-android-sdk/commit/e174127eb23e069c38373f545c52bab8c8615b87#diff-76992bc322b65ab94f299c05ba3cbd8b3ab2e32ebfdde6db5adba70fa9f5a26aR121) for a couple months now.

__Reviewer:__ @CodaFi

### CODE REVIEW

File: stdlib/public/Platform/SwiftGlibc.h.gyb [hunk 1]
```diff
@@ -27,8 +27,8 @@ headers = [
   'string.h',
   'tgmath.h',
   'time.h',
+  'util.h',
   'utmp.h',
-  'utmpx.h',
```
Comment (buttaface) [inline stdlib/public/Platform/SwiftGlibc.h.gyb:31]: This is duplicated on line 97 below, so removed this one.

**Comment by buttaface** [comment]
@CodaFi, if you approve and run the CI, I think we can get this in.

**Comment by CodaFi** [comment]
Nominated as rdar://89699689

@swift-ci test

**Comment by buttaface** [comment]
@CodaFi, been a week, tough to get the 5.6 branch approval?

**Comment by buttaface** [comment]
@CodaFi, I was waiting for a merge window to be announced for 5.6.1, but it never was and that just shipped. Can we get this in now?

**Comment by buttaface** [comment]
@airspeedswift, I think this qualifies as a bugfix for these alternate platforms, as this fixes several C++ Interop tests from the compiler validation suite. Let me know if we can get it into 5.6, or if I should close the pull.

**Comment by buttaface** [comment]
@airspeedswift, this only affects Android and OpenBSD, with the exception of the headers added to `stdlib/public/Platform/SwiftGlibc.h.gyb`. Please let me know what you think.

**Review by CodaFi — APPROVED** [review]

**Comment by buttaface** [comment]
The main reason is to bring these two alternative platforms in line with linux and other "SwiftGlibc" platforms, by also making this change that was done primarily for C++ interoperability with the Swift 5.5 release last year. There are some subtle issues with libc module scoping that this change also makes- see apple/swift-nio#2026 for an example- and this move will make sure cross-platform code doesn't hit such irregularities.

It's more a matter of aligning all platforms to a single-header modulemap, with this pull that is almost entirely scoped to these two less-used platforms.

**Comment by buttaface** [comment]
@shahmishal, let me know what you think of this one too.

**Comment by airspeedswift** [comment]
I’m still not seeing the reason why it’s key to get this into the release toolchain, instead of letting it happen in 5.7 naturally. C++ interop is not yet an official feature so nothing done in support of it should need to be back ported to the release. The 5.6.3 release is aimed at fixing bugs people are hitting when using the production toolchain, rather than helping facilitate WIP. Additionally, the linked PR to NIO implies that this change will cause source breaks – which is fine if it’s the right thing to do, but probably not in a patch release.

**Comment by buttaface** [comment]
> The 5.6.3 release is aimed at fixing bugs people are hitting when using the production toolchain, rather than helping facilitate WIP.

Right, so even ignoring C++ interop this pull removes that libc incompatibility between Linux and Android/OpenBSD, which has now been there since Swift 5.5. 

> Additionally, the linked PR to NIO implies that this change will cause source breaks – which is fine if it’s the right thing to do, but probably not in a patch release.

To be fair, I submitted this five months ago now, a month before the 5.6 release. I never said it was a "key" change, merely an attempt to regularize these niche Swift platforms of Android and OpenBSD.

If it "breaks" anything, it will be on those niche platforms, not on any of the major Swift platforms, unless you're worried about the two mentioned header additions for OpenBSD.

I realize it's a judgement call whether to merge this: my judgement is that since it primarily only affects Android and OpenBSD, we might as well get it in. If your judgement is otherwise, feel free to close.

**Comment by buttaface** [comment]
Ping, @airspeedswift, just need a decision on this before 5.6.3 is tagged. I think this is very low-risk and worth getting in, but if you don't think it's worth it, feel free to close.

**Comment by buttaface** [comment]
@shahmishal, no word from Ben in the last three weeks, can this still make the patch release?

**Comment by buttaface** [comment]
5.7 is out, so closing.

**Pull request closed without merging**